### PR TITLE
[WIP] Minimap

### LINKED ISSUE #111: Implement minimap
Should we limit the world terrain size for this? What if the world grows? The minimap scale changes accordingly to accomodate the whole world?

attached screenshot for reference
minimap
## <bountysource-plugin>

Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/5766131-implement-minimap?utm_campaign=plugin&utm_content=tracker%2F371436&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F371436&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by jfeo**
Here's my progress on a minimap.

There's probably a lot of style issues, etc. and I need to rebase some commits...

Right now, I think it's an issue that the Minimap registers itself with the engine for drawing, but then is deleted when the game is ended (dangling pointer?)

A good way to draw resources is also needed. Where are the colors/palette indices for them?

[Video](https://vid.me/Xx6B) for those who just want to see what it looks like.
implements and fixes #111

### CODE REVIEW

**Comment by Jon0** [comment]
The game_main class should only include pure game state data, so its not the best owner for the minimap, I would suggest looking at main.cpp, line 42-44 the where the are other components, and similarly minimap can be made on the stack (no 'new' constructor), 

You would need to change the constructor parameters to `minimap(Engine *engine, ... size, ... hudpos)` The minimap wont need to have pointers to the unit container and terrain since these can be retrieved in the `minimap::on_drawhud()` call using `GameMain *game = engine.get_game()` note: that will return nullptr when there is no active game, in which case no minimap should be drawn.

if the result is not null you will just use `game->terrain`, and `game->placed_units`, as before.

**Comment by franciscod** [comment]
oh the video looks awesome! nice work :D 

I almost forgot about gaben, and chuckled when he popped in the background :smile:

**Comment by TheJJ** [comment]
If you perform your rebase, you might want to use `git rebase -m -i upstream/master`.

**Comment by jfeo** [comment]
Is there any reason for me to keep the commit history of my own commits, or would it be okay to just squash it all into 1 commit? I think it would be easier for me

**Comment by TheJJ** [comment]
Sure, you can squash them into one if you like. Most of the time my work builds up incrementally so it makes sense keeping the history (e.g. `"implemented ..."`, then `"engine: make use of ..."`, then `"bla: enable ... globally"`) so each step could be reverted easily some day.

**Comment by jfeo** [comment]
Yes, I think I need to use a more structured approach in the future, instead of haphazardly adding everything and commiting :-P I'll just squash everything then.

File: libopenage/CMakeLists.txt [hunk 1]
```diff
@@ -23,7 +23,7 @@ add_sources(libopenage
 	player.cpp
 	screenshot.cpp
 	texture.cpp
-
+  minimap.cpp
```
Comment (franciscod) [inline libopenage/CMakeLists.txt]: whtitespace

File: libopenage/game_spec.cpp [hunk 2]
```diff
@@ -345,6 +348,9 @@ void GameSpec::load_terrain(AssetManager &am) {
 		// TODO: terrain double-define check?
 		terrain_data.terrain_id_priority_map[terrain_id]  = line->blend_priority;
 		terrain_data.terrain_id_blendmode_map[terrain_id] = line->blend_mode;
+    terrain_data.terrain_id_map_color_hi_map[terrain_id] = line->map_color_hi;
+    terrain_data.terrain_id_map_color_med_map[terrain_id] = line->map_color_med;
+    terrain_data.terrain_id_map_color_low_map[terrain_id] = line->map_color_low;
```
Comment (franciscod) [inline libopenage/game_spec.cpp]: whitespace

File: assets/shaders/minimap.vert.glsl [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+attribute vec4 vertex_position;
+attribute vec3 color;
+
+varying vec4 out_vertex_position;
+varying vec3 out_color;
+
+void main() {
+  out_vertex_position = vertex_position;
+  out_color = color;
+  gl_Position = gl_ModelViewProjectionMatrix * vertex_position;
```
Comment (TheJJ) [inline assets/shaders/minimap.vert.glsl]: hihi whitespace, tab indent pls

File: assets/shaders/minimap.frag.glsl [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+  ivec2 center = ivec2(minimap_size.x/2, minimap_size.y/2);
+  ivec2 vpos = minimap_orig + center - ivec2(out_vertex_position.x, out_vertex_position.y);
+  vec3 color;
+
+  if (!(abs(vpos.x) * center.y + abs(vpos.y) * center.x <= center.x * center.y))
+    discard;
+  else
+    color = out_color;
+
+  gl_FragColor = vec4(color, 1.0);
```
Comment (TheJJ) [inline assets/shaders/minimap.frag.glsl]: tab indent pls.

File: libopenage/terrain/terrain.h [hunk 5]
```diff
@@ -307,6 +311,22 @@ class Terrain {
...
+
+  /**
+   * get the red color component for a given terrain id.
+   */
+  uint8_t map_color_med(terrain_t terrain_id);
+
+  /**
+   * get the red color component for a given terrain id.
+   */
+  uint8_t map_color_low(terrain_t terrain_id);
```
Comment (TheJJ) [inline libopenage/terrain/terrain.h]: tab indent pls

**Comment by boombatower** [comment]
Does this predate the python interface and need to be updated for that approach or is this ready to be built on? Debating using this in referencing issue.

**Comment by jfeo** [comment]
Hey @boombatower. I'm a bit unsure of what you mean by the python interface. I just pulled from master locally, and it works, so there's that :-) In regards to #11, I suppose it would make sense to perhaps use this as a base (perhaps make it more generic?).
Arbitrary sizes are already possible, like this:
2015-09-10-193505_800x600_scrot

**Comment by TheJJ** [comment]
Used in #559 and integrated in the qt ui.

**Pull request closed without merging**